doc/design/platform-orchestrator-abstraction.md

```diff
@@ -0,0 +1,290 @@
+
+## Motivation
+
+The SR-IOV Network Operator has historically been tightly coupled to specific infrastructure platforms and Kubernetes distributions, particularly OpenShift. As the operator expanded to support different virtualization platforms like OpenStack, AWS, Oracle and various Kubernetes distributions, the need for a clean abstraction layer became apparent.
+
+### Use Cases
+
+1. **Multi-Platform Support**: Enable the operator to run efficiently on different infrastructure platforms (bare metal, OpenStack, AWS,Oracle, etc.) with platform-specific optimizations
+2. **Multi-Orchestrator Support**: Support different Kubernetes distributions (vanilla Kubernetes, OpenShift, etc.) with orchestrator-specific behaviors
+3. **Extensibility**: Make it easy to add new platforms and orchestrators without modifying core operator logic
+4. **Testing**: Enable better unit testing with mockable interfaces for platform and orchestrator specific operations
+
+### Goals
+
+* Create a clean abstraction layer that separates platform-specific logic from orchestrator-specific logic
+* Implement support for bare metal and OpenStack platforms
+* Implement support for Kubernetes and OpenShift orchestrators
+* Provide a plugin architecture that makes it easy to add new platforms and orchestrators
+* Maintain backward compatibility with existing functionality
+* Enable better testability through interface-based design
+
+### Non-Goals
+
+* Support all possible infrastructure platforms in the initial implementation
+* Change existing API structures or user-facing interfaces
+
+## Proposal
+
+### Workflow Description
+
+The operator will use two main abstraction layers:
+
+1. **Platform Interface**: Handles infrastructure-specific operations like device discovery, bridge management, and plugin selection
+2. **Orchestrator Interface**: Handles Kubernetes distribution-specific operations like cluster type detection, additional node draining logic, and cluster-specific configurations
+
+The platform is detected at startup based on node metadata and environment variables, while the orchestrator is detected based on cluster characteristics and available APIs.
+
+### API Extensions
+
+#### Platform Interface
+
+```golang
+type Interface interface {
+    Init() error
+    GetHostHelpers() helper.HostHelpersInterface
+    
+    DiscoverSriovDevices() ([]sriovnetworkv1.InterfaceExt, error)
+    DiscoverBridges() (sriovnetworkv1.Bridges, error)
+    
+    GetPlugins(ns *sriovnetworkv1.SriovNetworkNodeState) (plugin.VendorPlugin, []plugin.VendorPlugin, error)
+    SystemdGetPlugin(phase string) (plugin.VendorPlugin, error)
+}
+```
+
+#### Orchestrator Interface
+
+```golang
+type Interface interface {
+    ClusterType() consts.ClusterType
+    Flavor() consts.ClusterFlavor
+    BeforeDrainNode(context.Context, *corev1.Node) (bool, error)
+    AfterCompleteDrainNode(context.Context, *corev1.Node) (bool, error)
```

Comment:
nit: to be symmetric  with `BeforeDrainNode` lets use `AfterDrainNode`

Reply:
the problem with `AfterDrainNode` that it shound like after the operator finish to drain a node.

but that case is for after the operator finish to release a node from a drain(uncordon)